components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/FromInstanceTypeDeterminer.scala

```diff
@@ -35,15 +42,25 @@ object FromInstanceTypeDeterminer {
       case other =>
         Typed(other.getClass) match {
           case typedClass: TypedClass =>
+            val adjustedType = replaceWithGenericTypeIfNeeded(typedClass)
             ToJsonEncoder.default.encode(other) match {
-              case Valid(_)   => TypedObjectWithValue(typedClass, other)
-              case Invalid(_) => typedClass
+              case Valid(_)   => TypedObjectWithValue(adjustedType, other)
+              case Invalid(_) => adjustedType
```

Comment:
If we're just interested in boolean if its valid, ignoring the content of Valid or Invalid, we can do simpler:
```scala
            if (ToJsonEncoder.default.encode(other).isValid)
              TypedObjectWithValue(adjustedType, other)
            else
              adjustedType
```